SalesforceCommerceCloud/pwa-kit
# Fix: Handle non ASCII values in headers (@W-16087847@)

# Description

Related issue: https://github.com/SalesforceCommerceCloud/pwa-kit/issues/1739

This PR adds support for encoding non ASCII HTTP headers if `ENCODE_NON_ASCII_HTTP_HEADERS` is set to true in `packages/template-retail-react-app/app/constants.js`

Any headers with non ASCII values will be encoded using `encodedURIComponent` and an additional header will be passed, `x-encoded-headers` containing a comma separated list of the keys of headers that have been encoded.



# Types of Changes




> Breaking changes include:
>
> - Removing a public function or component or prop
> - Adding a required argument to a function
> - Changing the data type of a function parameter or return value
> - Adding a new peer dependency to `package.json`

# Changes

- (change1)

# How to Test-Drive This PR


## For request headers
1. Start tail logging: `npx @salesforce/pwa-kit-dev@latest tail-logs --project scaffold-pwa --environment joel-test-env`
2. Run the curl command: `curl -D- -so/dev/null 'https://scaffold-pwa-joel-test-env.mobify-storefront.com/' -H 'cf-region: Île-de-France' -H 'x-mobify-cachebreaker: 1'`
3. Observe 200 and not 500
5. Search for `@@@ REQUEST HEADERS BEFORE` in the tail logging
6. Observe `Cf-Region': 'ￃﾎle-de-France` (NOTE: I believe that AWS Lamda does some encoding/re-encoding under the hood where non ASCII characters get re-encoded into a different value unforuntately, I'll be looking to add an example on how to get around this)
7. Search for `@@@ REQUEST HEADERS BEFORE` in the tail logging
8. Observe `'Cf-Region': '%EF%BF%83%EF%BE%8Ele-de-France'` (`%EF%BF%83%EF%BE%8E` decoded is `ￃﾎ`) and `'x-encoded-headers': 'Cf-Region'`

## For response headers

In the deployed bundle, I've added some middleware to set a header with a non ASCII value for manual testing:

```js
        // test response header with this
        app.use((req, res, next) => {
            res.locals.nonAsciiHeader = 'テスト'
            // this throws an error
            // res.set('x-non-ascii-header', 'テスト')

            res.setHeader('x-non-ascii-header', res.locals.nonAsciiHeader);
            next()
        })
```

1. Make the same curl request as in the previous section: `curl -D- -so/dev/null 'https://scaffold-pwa-joel-test-env.mobify-storefront.com/' -H 'cf-region: Île-de-France' -H 'x-mobify-cachebreaker: 1'`
6. Observe 200 and that `x-non-ascii-header` is encoded and that `x-encoded-headers` is present with the encoded header key
`x-non-ascii-header: %E3%83%86%E3%82%B9%E3%83%88`
`x-encoded-headers: x-non-ascii-header`

# Checklists




## General


## Accessibility Compliance

You must check off all items in **one** of the follow two lists:


_or..._


## Localization
